### PR TITLE
Namespaced Map

### PR DESCRIPTION
**Pull request by naomijub**
`:abc{:a :b :c :d}`

### CODE REVIEW

File: examples/async.rs [hunk 1]
```diff
@@ -9,7 +9,7 @@ async fn foo() -> impl Future<Output = Edn> + Send {
 async fn main() {
     let edn = foo().await.await;
 
-    println!("{}", edn.to_string());
+    assert_eq!("[1, 1.5, \"hello\", :key, ]", edn.to_string());
```
Comment (evaporei) [inline examples/async.rs:12]: Should this be here? :thinking:

File: src/deserialize/mod.rs [hunk 2]
```diff
@@ -289,4 +289,20 @@ mod test {
             )]))
         )
     }
+
+    #[test]
+    fn namespaced_maps() {
+        let edn = ":abc{ 0 :val 1 :value}";
```
Comment (evaporei) [inline src/deserialize/mod.rs:295]: Have you tested with a space between `c` and `{`?

File: examples/async.rs (hunk 1 above)
Comment (naomijub) [reply examples/async.rs:12]: it was a failing test

File: src/deserialize/parse.rs [hunk 3]
```diff
@@ -30,11 +30,19 @@ pub(crate) fn parse_edn(c: Option<char>, chars: &mut std::str::Chars) -> Result<
     }
 }
 
-fn read_key(chars: &mut std::str::Chars) -> Edn {
-    let c_len = chars
+fn read_key_or_nsmap(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    let mut key_chars = chars
         .clone()
```
Comment (evaporei) [inline src/deserialize/parse.rs:35]: Why is this clone here again? I mean, there's a new one down there :thinking:

File: src/deserialize/mod.rs (hunk 2 above)
Comment (naomijub) [reply src/deserialize/mod.rs:295]: It cannot have the space, checked

File: src/deserialize/parse.rs (hunk 3 above)
Comment (naomijub) [reply src/deserialize/parse.rs:35]: because `take_while` takes the value it matched wrongly

Comment (naomijub) [reply src/deserialize/parse.rs:35]: All take_whiles are inside clones

File: src/deserialize/mod.rs (hunk 2 above)
Comment (evaporei) [reply src/deserialize/mod.rs:295]: But what does the spec say? I know we don't verify and return errors, but we should accept every valid EDN.

File: src/deserialize/parse.rs (hunk 3 above)
Comment (evaporei) [reply src/deserialize/parse.rs:35]: Yeah, but there's a new one on the function now, can we remove one of those, or is it necessary?

Comment (naomijub) [reply src/deserialize/parse.rs:35]: it is necessary, and we removed from here 
```rust
fn read_key(chars: &mut std::str::Chars, c_len: usize) -> Edn {
    let mut key = String::from(":");
    let key_chars = chars.take(c_len).collect::<String>();
    key.push_str(&key_chars);
    Edn::Key(key)
}
```

File: src/deserialize/mod.rs (hunk 2 above)
Comment (naomijub) [reply src/deserialize/mod.rs:295]: I checked the spec

Comment (naomijub) [reply src/deserialize/mod.rs:295]: What could happen is this `#myapp/Person {:first "Fred" :last "Mertz"}`. It is a different case, tagged map, and it is not really a valid edn, is similar to `#inst` and `#uuid` but needs a reader implemented in the client

Comment (evaporei) [reply src/deserialize/mod.rs:295]: oh shoot. Okay then

yeah, it is way more featured than JSON.

**Review by evaporei — APPROVED** [review]

**Pull request merged**